apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,427 @@
+		if err != nil {
+			return fmt.Errorf("failed to convert PublisherMerkleRoot: %w", err)
+		}
+		var publisherMerkleRoot [32]uint8
+		copy(publisherMerkleRoot[:], publisherMerkleRootBytes)
+
+		valueComputeAlgHashBytes, err := hexStringToByteArray(priceUpdate.StorkSignedPrice.StorkCalculationAlg.Checksum)
+		if err != nil {
+			return fmt.Errorf("failed to convert ValueComputeAlgHash: %w", err)
+		}
+
+		var valueComputeAlgHash [32]uint8
+		copy(valueComputeAlgHash[:], valueComputeAlgHashBytes)
+
+		rBytes, err := hexStringToByteArray(priceUpdate.StorkSignedPrice.TimestampedSignature.Signature.R)
+		if err != nil {
+			return fmt.Errorf("failed to convert R: %w", err)
+		}
+		var r [32]uint8
+		copy(r[:], rBytes)
+
+		sBytes, err := hexStringToByteArray(priceUpdate.StorkSignedPrice.TimestampedSignature.Signature.S)
+		if err != nil {
+			return fmt.Errorf("failed to convert S: %w", err)
+		}
+		var s [32]uint8
+		copy(s[:], sBytes)
+
+		vBytes, err := hexStringToByteArray(priceUpdate.StorkSignedPrice.TimestampedSignature.Signature.V)
+		if err != nil {
+			return fmt.Errorf("failed to convert V: %w", err)
+		}
+		v := uint8(vBytes[0])
+
+		updateData := contract.TemporalNumericValueEvmInput{
+			TemporalNumericValue: contract.TemporalNumericValue{
+				TimestampNs:    uint64(priceUpdate.StorkSignedPrice.TimestampedSignature.Timestamp),
+				QuantizedValue: quantizedPrice,
+			},
+			Id:                  assetId,
+			PublisherMerkleRoot: publisherMerkleRoot,
+			ValueComputeAlgHash: valueComputeAlgHash,
+			R:                   r,
+			S:                   s,
+			V:                   v,
+			TreasuryId:          treasuryId,
+		}
+
+		instruction, err := contract.NewUpdateTemporalNumericValueEvmInstruction(
+			updateData,
+			sci.configAccount,
+			treasuryAccount,
+			feedAccount,
+			sci.payer.PublicKey(),
+			solana.SystemProgramID,
+		).ValidateAndBuild()
+		if err != nil {
+			return fmt.Errorf("failed to build instruction: %w", err)
+		}
+		instructions = append(instructions, instruction)
+
+	}
+
+	recentBlockHash, err := sci.client.GetLatestBlockhash(context.Background(), rpc.CommitmentFinalized)
+
+	if err != nil {
+		return fmt.Errorf("failed to get recent blockhash: %w", err)
+	}
+
+	tx, err := solana.NewTransaction(
+		instructions,
+		recentBlockHash.Value.Blockhash,
+		solana.TransactionPayer(sci.payer.PublicKey()),
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create transaction: %w", err)
+	}
+
+	_, err = tx.Sign(
+		func(key solana.PublicKey) *solana.PrivateKey {
+			if key == sci.payer.PublicKey() {
+				return &sci.payer
+			}
+			return nil
+		})
+	if err != nil {
+		return fmt.Errorf("failed to sign transaction: %w", err)
+	}
+
+	sig, err := sci.client.SendTransaction(context.Background(), tx)
+	if err != nil {
+		return fmt.Errorf("failed to send transaction: %w", err)
+	}
+	// check for confirmation without blocking
+	go sci.confirmTransaction(sig)
```

Comment:
This might kick off a ton of go routines. It would probably be better to throw this onto a channel and use a thread pool in a separate go routine to do these confirmations.

Reply:
Changed. I feel very strongly that confirmation should never block updates, so I moved to a limited worker pool (10 go routines for now, but no strong opinions on this number) to limit cpu overhead while using an unbounded queue to prevent blocking, which a normal buffered channel might do. After the inital lag time for the first confirmations to hit, the (avg) confirmation rate then keeps pace with the submission rate in my testing, so actual unbounded growth is unlikely, however imo preferable to blocking. Let me know your thoughts.

Reply:
Its worth noting that we can actually calculate the maximum number of necessary workers with:

$$
\text{workers} = \frac{txs}{block}
$$

expanded:

$$
\text{workers} = \frac{\lceil \frac{\text{assets}}{4} \rceil}{\text{batchwindow}} \times \text{blocktime}
$$ 

where:

$$
\begin{align}
&\text{assets} = \text{number of assets} \\
&\text{batchwindow} = \text{batching window in seconds} \\
&\text{blocktime} = \text{average solana blocktime in seconds/block}
\end{align}
$$

This of course assumes every asset gets an update with great enough change in every batching window, which from my testing so seems not to be the case. In any case, not sure how complicated we want to make it, or what reasonable upper bounds are for the number of assets one pusher will ever be running for, But I can implement this math at interactor initialization so at least the number of workers is not a bit arbitrary like it is now.